src/dotty/tools/dotc/transform/PostTyper.scala

```diff
@@ -97,17 +97,37 @@ class PostTyper extends MacroTransform with IdentityDenotTransformer  { thisTran
    *  Revisit this issue once we have implemented `inline`. Then we can demand
    *  purity of the prefix unless the selection goes to an inline val.
    */
-  private def normalizeTree(tree: Tree)(implicit ctx: Context): Tree = tree match {
-    case tree: TypeTree => tree
-    case _ =>
-      if (tree.isType) {
-        Checking.boundsChecker.traverse(tree)
-        TypeTree(tree.tpe).withPos(tree.pos)
-      }
-      else tree.tpe.widenTermRefExpr match {
-        case ConstantType(value) if isIdempotentExpr(tree) => Literal(value)
-        case _ => tree
-      }
+  private def normalizeTree(tree: Tree)(implicit ctx: Context): Tree = {
+    def literalize(tp: Type): Tree = tp.widenTermRefExpr match {
+      case ConstantType(value) if isIdempotentExpr(tree) => Literal(value)
+      case _ => tree
+    }
+    def norm(tree: Tree) =
+      if (tree.isType) TypeTree(tree.tpe).withPos(tree.pos)
+      else literalize(tree.tpe)
+    tree match {
+      case tree: TypeTree =>
+        tree
+      case AppliedTypeTree(tycon, args) =>
+        // If `args` is a list of named argiments, return corresponding type parameters,
+        // or abstract types otherwise return type parameters unchanged
+        def matchNamed(tparams: List[TypeSymbol], args: List[Tree]): List[Symbol] = args match {
+          case (_: NamedArg) :: _ =>
```

Comment:
This pattern match would be simpler if it was done on `args.head` instead of `args`